Async instantiating and unboxing thunks

MichalStrehovsky opened the pull request:
We need to:

* Compile the unboxing/instantiating thunk as `CORJIT_FLAG_ASYNC` when the target is AsyncCall
* Use the TailAwait intrinsic to just return the suspended state from the called method instead of trying to suspend the instantiating thunk too.

Also includes a JIT change from #124175.

Fixes #121781.

Cc @dotnet/ilc-contrib @jakobbotsch

MichalStrehovsky committed:
Async instantiating and unboxing thunks

VSadov commented:
> We do not have that right now.

Maybe you can hack it by emitting something like:

```cs
// < do the inner call>

// fetch the callee`s continuation, if any
Continuation c = AsyncCallContinuation();
if (c != null)
{
      // return calee's continuation to the caller.
      AsyncSuspend(c);
      // UNREACHABLE	
}

return result;
```

This has not been tried, so it is a big YMMV. 
Also, I think you would still need to mark the unboxing stub as AsyncCall, but you may get away with not providing a resume.

jakobbotsch committed:
JIT: Add an AsyncHelpers.TailAwait intrinsic

Can be used to instruct the JIT that an upcoming await should not create
an explicit suspension point. This can be used to optimize
instantiating/unboxing stubs and will simplify the NAOT side of the
implementation of those.

jakobbotsch committed:
Typo

jakobbotsch committed:
Nit

jakobbotsch committed:
Fix x86 shared return

MichalStrehovsky committed:
Report as async state machine and use TailAwait

MichalStrehovsky committed:
Do not use in the VM for now

jakobbotsch committed:
Copilot feedback

jakobbotsch reviewed (approved):
LGTM, but @EgorBo can you PTAL on the JIT parts too?
